### PR TITLE
Add summation algorithms

### PR DESCRIPTION
**Pull request by b3liever**
Please pick one:
```nim
func sumKbn*[T: SomeFloat](v: Indexable[T]): T =
   # Kahan (compensated) summation: O(1) error growth, at the expense
   # of a considerable increase in computational expense.
   if len(v) == 0: return
   var sum = v[0]
   var c = T(0)
   for i in 1 ..< len(v):
      let vi = v[i]
      let t = sum + vi
      if abs(sum) >= abs(vi):
         c += (sum - t) + vi
      else:
         c += (vi - t) + sum
      sum = t
   result = sum + c

func sumPairwise[T](v: Indexable[T], i0, n: int): T =
   if n < 128:
      result = v[i0]
      for i in i0 + 1 ..< i0 + n:
         result += v[i]
   else:
      let n2 = n div 2
      result = sumPairwise(v, i0, n2) + sumPairwise(v, i0 + n2, n - n2)

func sumPairs*[T: SomeFloat](v: Indexable[T]): T =
   # Pairwise (cascade) summation of v[i0:i0+n-1], which O(log n) error growth
   # [vs O(n) for a simple loop] with negligible performance cost if
   # the base case is large enough.  See, e.g.:
   #        http://en.wikipedia.org/wiki/Pairwise_summation
   #        Higham, Nicholas J. (1993), "The accuracy of floating point
   #        summation", SIAM Journal on Scientific Computing 14 (4): 783–799.
   # In fact, the root-mean-square error growth, assuming random roundoff
   # errors, is only O(sqrt(log n)), which is nearly indistinguishable from O(1)
   # in practice.  See:
   #        Manfred Tasche and Hansmartin Zeuner, Handbook of
   #        Analytic-Computational Methods in Applied Mathematics (2000).
   let n = len(v)
   if n == 0: T(0) else: sumPairwise(v, 0, n)

func msum*[T: SomeFloat](v: Iterable[T]): T =
   ## Full precision sum of values in iterable. Returns the value of the
   ## sum, rounded to the nearest representable floating-point number
   ## using the round-half-to-even rule
   var partials = newSeq[T]()
   for x in v:
      var x = x
      var i = 0
      for y in partials:
         var y = y
         if abs(x) < abs(y):
            swap(x, y)
         let hi = x + y
         let lo = y - (hi - x)
         if lo != T(0):
            partials[i] = lo
            i.inc
         x = hi
      setLen(partials, i + 1)
      partials[i] = x
   result = sum(partials)
```

### CODE REVIEW

**Comment by b3liever** [comment]
isClose is a very usefull function! but i agree it isn't in the scope of this PR and I will remove it completely.

**Comment by mratsim** [comment]
Keep it for testing, so you can make it private.

**Comment by Araq** [comment]
"Pairwise (cascade) summation" seems much better.

**Comment by b3liever** [comment]
Can I keep `sumKbn`? If `n < 128` `sumPairwise` just does sum.

**Comment by LemonBoy** [comment]
Show some numbers please (you can use [this](https://github.com/LemonBoy/criterion.nim) if you want)

**Comment by b3liever** [comment]
@LemonBoy I dont know how to make this work but I tried:
```nim
let cfg = newDefaultConfig()

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc sumKbn1(input: seq[float]) {.measure: tc1.} =
      discard sumKbn(input) # == 7.4854708605503451

   proc sumKbn2(input: seq[float]) {.measure: tc2.} =
      discard sumKbn(input) # == -0.69264743055982025

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc sumPairs1(input: seq[float]) {.measure: tc1.} =
      discard sumPairs(input) # == 7.4854708605503451

   proc sumPairs2(input: seq[float]) {.measure: tc2.} =
      discard sumPairs(input) # == -0.69264743055982025

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc msum1(input: seq[float]) {.measure: tc1.} =
      discard msum(input) # == 7.4854708605503451

   proc msum2(input: seq[float]) {.measure: tc2.} =
      discard msum(input) # == -0.69264743055982025
```

**Comment by LemonBoy** [comment]
I've pushed some updates to criterion.nim and did two small modifications to your code:
- Use `blackBox` instead of `discard` in order to prevent the compiler from optimizing the unused value away
- Use `{.measure: [tc1].}`, otherwise it'll run a separate benchmark for each of the elements in the seq.

Code:
<details>

```nim
import criterion
import math

let cfg = newDefaultConfig()

func sumKbn*[T: SomeFloat](v: openArray[T]): T =
   # Kahan (compensated) summation: O(1) error growth, at the expense
   # of a considerable increase in computational expense.
   if len(v) == 0: return
   var sum = v[0]
   var c = T(0)
   for i in 1 ..< len(v):
      let vi = v[i]
      let t = sum + vi
      if abs(sum) >= abs(vi):
         c += (sum - t) + vi
      else:
         c += (vi - t) + sum
      sum = t
   result = sum + c

func sumPairwise[T](v: openArray[T], i0, n: int): T =
   if n < 128:
      result = v[i0]
      for i in i0 + 1 ..< i0 + n:
         result += v[i]
   else:
      let n2 = n div 2
      result = sumPairwise(v, i0, n2) + sumPairwise(v, i0 + n2, n - n2)

func sumPairs*[T: SomeFloat](v: openArray[T]): T =
   # Pairwise (cascade) summation of v[i0:i0+n-1], which O(log n) error growth
   # [vs O(n) for a simple loop] with negligible performance cost if
   # the base case is large enough.  See, e.g.:
   #        http://en.wikipedia.org/wiki/Pairwise_summation
   #        Higham, Nicholas J. (1993), "The accuracy of floating point
   #        summation", SIAM Journal on Scientific Computing 14 (4): 783–799.
   # In fact, the root-mean-square error growth, assuming random roundoff
   # errors, is only O(sqrt(log n)), which is nearly indistinguishable from O(1)
   # in practice.  See:
   #        Manfred Tasche and Hansmartin Zeuner, Handbook of
   #        Analytic-Computational Methods in Applied Mathematics (2000).
   let n = len(v)
   if n == 0: T(0) else: sumPairwise(v, 0, n)

func msum*[T: SomeFloat](v: openArray[T]): T =
   ## Full precision sum of values in iterable. Returns the value of the
   ## sum, rounded to the nearest representable floating-point number
   ## using the round-half-to-even rule
   var partials = newSeq[T]()
   for x in v:
      var x = x
      var i = 0
      for y in partials:
         var y = y
         if abs(x) < abs(y):
            swap(x, y)
         let hi = x + y
         let lo = y - (hi - x)
         if lo != T(0):
            partials[i] = lo
            i.inc
         x = hi
      setLen(partials, i + 1)
      partials[i] = x
   result = sum(partials)

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc sumKbn1(input: seq[float]) {.measure: [tc1].} =
      blackBox sumKbn(input) # == 7.4854708605503451

   proc sumKbn2(input: seq[float]) {.measure: [tc2].} =
      blackBox sumKbn(input) # == -0.69264743055982025

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc sumPairs1(input: seq[float]) {.measure: [tc1].} =
      blackBox sumPairs(input) # == 7.4854708605503451

   proc sumPairs2(input: seq[float]) {.measure: [tc2].} =
      blackBox sumPairs(input) # == -0.69264743055982025

benchmark(cfg):
   # test case 1
   var tc1: seq[float]
   for n in 1 .. 1000:
      tc1.add 1.0 / n.float
   # test case 2
   var tc2: seq[float]
   for n in 1 .. 1000:
      tc2.add pow(-1.0, n.float) / n.float

   proc msum1(input: seq[float]) {.measure: [tc1].} =
      blackBox msum(input) # == 7.4854708605503451

   proc msum2(input: seq[float]) {.measure: [tc2].} =
      blackBox msum(input) # == -0.69264743055982025
```
</details>

**Comment by b3liever** [comment]
<details>
<summary>First test</summary>
<br>
```
Benchmark: sumKbn1(input = seq[float])
Collected 188 samples
Warning: Found 16 mild and 1 extreme outliers in the time measurements
Warning: Found 16 mild and 1 extreme outliers in the cycles measurements
Time
  Time: 3.2070us (3.1957us .. 3.2240us)
  R²:   0.9999 (0.9996 .. 1.0000)
  Mean: 3.1817us (3.1557us .. 3.2080us)
  Std:  170.9578ns (145.1853ns .. 198.8687ns)
Cycles
  Cycles: 6'081cycles (6'059cycles .. 6'111cycles)
  R²:     0.9999 (0.9996 .. 1.0000)
  Mean:   6'028cycles (5'979cycles .. 6'073cycles)
  Std:    319cycles (267cycles .. 370cycles)

Benchmark: sumPairs1(input = seq[float])
Collected 198 samples
Warning: Found 19 mild and 2 extreme outliers in the time measurements
Warning: Found 18 mild and 2 extreme outliers in the cycles measurements
Time
  Time: 1.9573us (1.9432us .. 1.9772us)
  R²:   0.9992 (0.9987 .. 0.9995)
  Mean: 1.9496us (1.9342us .. 1.9642us)
  Std:  106.5782ns (89.2799ns .. 121.2878ns)
Cycles
  Cycles: 3'711cycles (3'685cycles .. 3'748cycles)
  R²:     0.9992 (0.9988 .. 0.9995)
  Mean:   3'692cycles (3'666cycles .. 3'720cycles)
  Std:    197cycles (169cycles .. 224cycles)

Benchmark: msum1(input = seq[float])
Collected 152 samples
Time
  Time: 18.7254us (18.6947us .. 18.7831us)
  R²:   0.9998 (0.9997 .. 0.9999)
  Mean: 20.2736us (19.9962us .. 20.5533us)
  Std:  1.7095us (1.5613us .. 1.8829us)
Cycles
  Cycles: 35'511cycles (35'451cycles .. 35'621cycles)
  R²:     0.9998 (0.9997 .. 0.9999)
  Mean:   38'439cycles (37'970cycles .. 38'943cycles)
  Std:    3'242cycles (2'948cycles .. 3'520cycles)
```
</details>
<details>
<summary>Second test</summary>
<br>
```
Benchmark: sumKbn2(input = seq[float])
Collected 187 samples
Warning: Found 8 mild and 37 extreme outliers in the time measurements
Warning: Found 6 mild and 38 extreme outliers in the cycles measurements
Time
  Time: 3.4176us (3.3579us .. 3.4951us)
  R²:   0.9983 (0.9970 .. 0.9994)
  Mean: 3.3499us (3.3317us .. 3.3686us)
  Std:  134.1876ns (110.5441ns .. 159.1739ns)
Cycles
  Cycles: 6'481cycles (6'367cycles .. 6'630cycles)
  R²:     0.9983 (0.9970 .. 0.9994)
  Mean:   6'347cycles (6'311cycles .. 6'384cycles)
  Std:    252cycles (203cycles .. 300cycles)

Benchmark: sumPairs2(input = seq[float])
Collected 198 samples
Warning: Found 22 mild and 34 extreme outliers in the time measurements
Warning: Found 24 mild and 31 extreme outliers in the cycles measurements
Time
  Time: 2.0542us (2.0309us .. 2.0910us)
  R²:   0.9986 (0.9970 .. 0.9995)
  Mean: 2.0114us (2.0021us .. 2.0209us)
  Std:  67.9016ns (53.2332ns .. 82.3600ns)
Cycles
  Cycles: 3'895cycles (3'848cycles .. 3'963cycles)
  R²:     0.9986 (0.9969 .. 0.9995)
  Mean:   3'809cycles (3'792cycles .. 3'827cycles)
  Std:    124cycles (93cycles .. 152cycles)

Benchmark: msum2(input = seq[float])
Collected 151 samples
Warning: Found 3 mild and 26 extreme outliers in the time measurements
Warning: Found 5 mild and 26 extreme outliers in the cycles measurements
Time
  Time: 20.4183us (20.3236us .. 20.5052us)
  R²:   0.9996 (0.9991 .. 0.9999)
  Mean: 19.9066us (19.8538us .. 19.9709us)
  Std:  373.3098ns (211.3809ns .. 549.9631ns)
Cycles
  Cycles: 38'722cycles (38'520cycles .. 38'885cycles)
  R²:     0.9996 (0.9991 .. 0.9999)
  Mean:   37'744cycles (37'646cycles .. 37'871cycles)
  Std:    708cycles (402cycles .. 1'074cycles)
```
</details>

**Comment by b3liever** [comment]
What you want me to do to close this PR?

File: lib/pure/math.nim [hunk 1]
```diff
@@ -820,3 +860,24 @@ when isMainModule:
...
     doAssert log2(1.0'f32) == 0.0'f32
     doAssert classify(log2(0.0'f32)) == fcNegInf
+
+  block: # sumKbn, sumPairs
+    var epsilon = 1.0
+    while 1.0 + epsilon != 1.0:
+      epsilon /= 2.0
+    let data = @[1.0, epsilon, -epsilon]
+    assert sumKbn(data) == 1.0
+    assert sumPairs(tc0) != 1.0 # known to fail
```
Comment (Araq) [inline lib/pure/math.nim]: Undeclared identifier: ``tc0``

**Comment by b3liever** [comment]
I dont know what to say.

**Comment by bluenote10** [comment]
> I don't want this in math. It seems to be used rarely and math is a very commonly imported library.

Why is this all decided by you now instead of considering the opinion of other maintainers or even the community?

Since the functions are generic they don't increase code size; compilation overhead should be minimal; added complexity/maintainability seems reasonable, and it is a common feature for a math standard library. To me this would have been a useful addition and I already had plans to use it.

**Comment by krux02** [comment]
well maybe my reaction was a bit harsh. @b3liever I am sorry for my comment. The problem is, things cannot really ever be taken out if they ever got introduced somewhere in the standard library. And library featuers that are not very very well thought through will haunt the maintainer for a very very long time. So I am not agains this feature per se, but if it is added, it should be added to it's own file in ``/lib/std``

**Comment by b3liever** [comment]
No problem here. Thing is, the math library could be better, and if you don't accept PR, nothing will change.

**Comment by krux02** [comment]
This was just my personal opinion, but that doesn't mean that it is the last word spoken. But my opinion stays the same, new features should be in their own file first. Only when they have been proven valuable enough they will be included in the math library.

**Comment by Araq** [comment]
I like it, if we put it under ``std / sums``

**Comment by b3liever** [comment]
Cool, I will do it, once internet connectivity is restored.

**Pull request merged**